### ISSUE TITLE
[1.19 regression] Can no longer use `types.NoneType()` as a condition in match/case

### ISSUE THREAD

**Issue by chadhpaine** [python/mypy#20367]
**Bug Report**

Prior to mypy 1.19, it was possible to use `types.NoneType()` as a pattern-matcher in match/case syntax for values whose 'revealed type' includes 'the None type'.

**To Reproduce**

```python
import re
import types

def fun(val: str | re.Pattern | None):
    match val:
        case str():
            print('string')
        case re.Pattern():
            print('pattern')
        case types.NoneType():  # main.py:10: error: Expected type in class pattern; found "type[None]"  [misc]
            print('None')

fun('123')
fun(re.compile('456'))
fun(None)
```

**Expected Behavior**

No error or warning is raised.

**Actual Behavior**

```
main.py:10: error: Expected type in class pattern; found "type[None]"  [misc]
Found 1 error in 1 file (checked 1 source file)
```

**Your Environment**

Reproduces on mypy-play.net.
https://mypy-play.net/?mypy=latest&python=3.12&gist=3dc5c105d72ba997443f1c4f58a179a0

- Mypy version used: 1.19.0
- Mypy command-line flags: n/a
- Mypy configuration options from `mypy.ini` (and other config files): n/a
- Python version used: 3.12

**Comment by hauntsaninja** [comment]
Thanks, bisects to https://github.com/python/mypy/pull/20042

**Comment by hauntsaninja** [comment]
Something like this would fix, but probably want something a little more general
```
diff --git a/mypy/checkpattern.py b/mypy/checkpattern.py
index 3c51c4106..19a99f4e4 100644
--- a/mypy/checkpattern.py
+++ b/mypy/checkpattern.py
@@ -556,6 +557,8 @@ class PatternChecker(PatternVisitor[PatternType]):
             fallback = self.chk.named_type("builtins.function")
             any_type = AnyType(TypeOfAny.unannotated)
             typ = callable_with_ellipsis(any_type, ret_type=any_type, fallback=fallback)
+        elif isinstance(p_typ, TypeType) and isinstance(p_typ.item, NoneType):
+            typ = p_typ.item
         elif not isinstance(p_typ, AnyType):
             self.msg.fail(
                 message_registry.CLASS_PATTERN_TYPE_REQUIRED.format(
```

**Comment by A5rocks** [comment]
I think that is the best we can do, unfortunately. `type_info` is a tempting variable to use, but doesn't work on type aliases.

Probably I should look at the logic in `ExpressionChecker` to see what other cases might be returned as a `TypeType`, but for now I'll just make a PR.
